packetbeat/protos/cassandra/parser.go

```diff
@@ -0,0 +1,246 @@
+		return err
+	}
+
+	for p.buf.Total() > 0 {
+		if p.message == nil {
+			// allocate new message object to be used by parser with current timestamp
+			p.message = p.newMessage(ts)
+		}
+
+		msg, err := p.parse()
+		if err != nil {
+			return err
+		}
+		if msg == nil {
+			break // wait for more data
+		}
+
+		// reset buffer and message -> handle next message in buffer
+		p.buf.Reset()
+		p.message = nil
+
+		// call message handler callback
+		if err := p.onMessage(msg); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *parser) newMessage(ts time.Time) *message {
+	return &message{
+		Message: applayer.Message{
+			Ts: ts,
+		},
+	}
+}
+
+func (p *parser) parse() (*message, error) {
+
+	// if p.frame is nil then create a new framer, or continue to process the last message
+	if p.framer == nil {
+		if isDebug {
+			logp.Debug("cassandra", "start new framer")
+		}
+		p.framer = NewFramer(&p.buf, p.config.compressor)
+	}
+
+	// check if the frame header were parsed or not
+	if p.framer.Header == nil {
+		if isDebug {
+			logp.Debug("cassandra", "start to parse header")
+		}
+		if !p.buf.Avail(9) {
+			logp.Err("not enough head bytes, ignore")
+			return nil, nil
+		}
+
+		_, err := p.framer.ReadHeader()
+		if err != nil {
+			logp.Err("%v", err)
+			p.framer = nil
+			return nil, err
+		}
+	}
+
+	msg := p.message
+	msg.data = make(map[string]interface{})
+
+	if p.framer.Header.BodyLength > 0 {
+
+		//let's wait for enough buf
+		if !p.buf.Avail(p.framer.Header.BodyLength) {
+			if isDebug {
+				logp.Debug("cassandra", " buf not enough for body, waiting for more, return")
+			}
+			return nil, nil
+		} else {
+			//check if the ops already ignored
+			if p.config.ignoredOps != nil && len(p.config.ignoredOps) > 0 {
+				if isDebug {
+					logp.Debug("cassandra", "ignoreOPS configed, let's check")
+				}
+				v := p.config.ignoredOps[p.framer.Header.Op.String()]
+				if v != nil {
+					if isDebug {
+						logp.Debug("cassandra", fmt.Sprintf("Ops: %s was marked to be ignored, ignoring, request:%v", p.framer.Header.Op.String(), p.framer.Header.Version.IsRequest()))
+					}
+					p.buf.Collect(p.framer.Header.BodyLength)
+
+					finalCollectedFrameLength := p.buf.BufferConsumed()
+					if finalCollectedFrameLength-p.framer.Header.HeadLength != p.framer.Header.BodyLength {
+						return nil, errors.New("data messed while parse frame body")
+					}
+					// as we already igore the content, we now mark the result is completed
+					p.message.ignored = true
+				}
+			}
+
+			// start to parse body
+			if !p.message.ignored {
+				if isDebug {
+					logp.Debug("cassandra", "start read frame")
+				}
+				data, err := p.framer.ReadFrame()
+				if err != nil {
+					// if the frame parsed failed, should ignore the whole message
+					p.framer = nil
+					return nil, err
+				}
+
+				msg.data = data
+
+				// dealing with un-parsed content
+				frameParsedLength := p.buf.BufferConsumed()
+
+				// collect leftover
```

Comment:
not padding, some message detail is less interested, i just parser the metadata, but the left content also need to read but won't process